scrubadub/scrubbers.py

```diff
@@ -263,33 +271,55 @@ def iter_filth_documents(
         if not isinstance(documents, (dict, list)):
             raise TypeError('documents must be one of a string, list of strings or dict of strings.')
 
-        if run_post_processors:
-            # Only collect the filts into a list if we need to do post processing
-            filth_list = []  # type: List[Filth]
+        # Figures out which detectors have iter_filth_documents and applies to them
+
+        document_detectors_names = []
+        filth_list = []
+
+        for name, detector in self._detectors.items():
+            document_iterator = getattr(detector, 'iter_filth_documents', None)
+            if callable(document_iterator):
+                document_detectors_names.append(name)
+                for filth in document_iterator(documents):
```

Comment:
Then we can re-use most of the stuff that was in iter_filth in here like this:

```
    def iter_filth(
            self, text: str, document_name: Optional[str] = None, run_post_processors: bool = True,
    ) -> Generator[Filth, None, None]:
        """Iterate over the different types of filth that can exist.
        """
        for filth in self.iter_filth_documents(
            documents={document_name: text}, run_post_processors=run_post_processors
        ):
            yield filth

    def iter_filth_documents(
            self,
            documents: Union[Sequence[str], Dict[Optional[str], str]],
            run_post_processors: bool = True
    ) -> Generator[Filth, None, None]:
        """Iterate over the different types of filth that can exist."""
        if not isinstance(documents, (dict, list)):
            raise TypeError('documents must be one of a string, list of strings or dict of strings.')

        if isinstance(documents, dict):
            document_names, document_texts = zip(*documents.items())
        elif isinstance(documents, (tuple, list)):
            document_texts = documents
            document_names = [str(x) for x in range(len(documents))]

        # currently doing this by aggregating all_filths and then sorting
        # inline instead of with a Filth.__cmp__ method, which is apparently
        # much slower http://stackoverflow.com/a/988728/564709
        #
        # NOTE: we could probably do this in a more efficient way by iterating
        # over all detectors simultaneously. just trying to get something
        # working right now and we can worry about efficiency later
        filth_list = []  # type: List[Filth]
        for name, detector in self._detectors.items():
            if callable(getattr(detector, 'iter_filth_documents', None)):
                for filth in detector.iter_filth_documents(document_names, document_texts):
                    if not isinstance(filth, Filth):
                        raise TypeError('iter_filth must always yield Filth')
                    filth_list.append(filth)
            else:
                for document_name, text in zip(document_names, document_texts):
                    for filth in detector.iter_filth(text, document_name=document_name):
                        if not isinstance(filth, Filth):
                            raise TypeError('iter_filth must always yield Filth')
                        filth_list.append(filth)

        # This is split up so that we only have to use lists if we have to post_process Filth
        if run_post_processors:
            all_filths = list(self._merge_filths(filth_list))
            all_filths = list(self._post_process_filth_list(all_filths))

            # Here we loop over a list of Filth...
            for filth in all_filths:
                yield filth
        else:
            # ... but here, we're using a generator. If we try to use the same variable it would have two types and
            # fail static typing in mypy
            for filth in self._merge_filths(filth_list):
                yield filth
```

The above doesn't pass mypy yet, but the test cases ran. Does this successfully simplify things a bit?

Reply:
That sounds good. I actually started implementing something along these lines, but since I didn't want to to mess around with the scrubber code that much, I decided to change my mind and just code "around" it. But your suggestion seems much better for organisation. I will give it a shot!